nn/transformer.py

```diff
@@ -0,0 +1,253 @@
+"""
+Transformer Modules
+"""
+
+from __future__ import annotations
+from typing import Optional, Any, Union, Callable
+from .. import nn
+
+
+class TransformerEncoderLayer(nn.Module):
+  """
+  Defines one layer of a standard transformer encoder
+  """
+  def __init__(self, dim_model: int, num_heads: int, dim_ff: int = 2048, dropout: float = 0.1,
+               activation: Callable[[nn.LayerRef], nn.LayerRef] = nn.relu, norm_eps: float = 1e-5,
+               norm_first: bool = False, norm=nn.layer_norm) -> None:
+    """
+    :param dim_model: hidden dim, PyTorch name: d_model
+    :param num_heads: number heads, PyTorch name: nhead
+    :param dim_ff: dimension of feedforward layer, PyTorch name: dim_feedforward
+    :param dropout: Dropout value, PyTorch name: dropout
+    :param activation: activation function
+    :param norm_eps: Epsilon value for layer normalization
+    :param norm_first: if ``True`` will perform normalization before other att and ff operations, otherwise after
+    :param norm: normalization function
+    """
+    super().__init__()
+    time_dim = nn.DimensionTag(nn.DimensionTag.Types.Spatial, description="time")
+    self.self_attn = nn.SelfAttention(
+      axis=time_dim, key_dim_total=dim_model, value_dim_total=dim_model, num_heads=num_heads, att_dropout=dropout)
```

Comment:
`att_dropout=dropout` is wrong.

Reply:
Oh wait, actually not? It uses the same dropout also for the attention dropout? At least in the PyTorch implementation. But is this good? Does it make sense to use the same dropout value?

Reply:
Yes I would expect it to be separate. I would not expect that the same dropout value makes sense for both attention weights and for other activations.

What do you usually use? Or what is usually used in the literature? Is it common that people use the same dropout for both?

If this is very common and a good choice, then we can leave it. Otherwise I think we should separate it.

Reply:
At least in PyTorch its the same value I think, looking at Attention is all you need I am not sure if they even used dropout there. At least the paper does not explicitly state it from what I can see. But if you want we can separate and use the same defaults.

Reply:
Look in actual configs from publications (ours, others) or production systems or our baselines.

It also doesn't make sense to use the same defaults. The defaults should be good. That was what I asked before. What are common good defaults here?

Reply:
From what I could find 0.1 is the default for both attention dropout and (general) dropout, correlating to the 6 Dec/Enc Layer Transformer. For other sizes there are other defaults, but if we use 6 as default this should be standard in literature. I separated it from standard dropout now.

Reply:
Where? Esp what recent papers use that? Please add such references to the docstring as well.

Reply:
But also, what do we usually use? What do you have in your baselines?

Reply:
espnet uses 0.1 for both for TTS, for ASR Encoder/Decoder they use 0.1 dropout, 0.0 self attention dropout
fairseq uses 0.1 for dropout and 0.0 for self attention dropout
Lingvo uses 0.0 as a default
[ASR Transformer](https://ieeexplore.ieee.org/stamp/stamp.jsp?tp=&arnumber=8462506) uses 0.1 for both
for some other papers I could not find information on what they use or they referenced to the above paper, like the espnet [study](https://merl.com/publications/docs/TR2019-158.pdf)

Reply:
So my suggestion (if we want to fill it and not leave it empty) would be to use 0.1 for residual dropout and either 0.1 or 0.0 for attention dropout. 0.0 would match with your default from the `SelfAttention` layer.

Reply:
Maybe then 0.1 for both. I think it is a good choice to use it always also for the att weights.

Also, `SelfAttention` maybe should change the default.